core/src/main/java/org/infinispan/stream/impl/DistributedCacheStream.java

```diff
@@ -513,9 +519,30 @@ private RehashIterator(Iterable<IntermediateOperation> intermediateOperations) {
       protected S getNext() {
          do {
             if (currentIterator == null) {
-               log.tracef("Creating rehash iterator for segments %s", segmentsToUse);
-               currentIterator = nonRehashRemoteIterator(segmentsToUse, receivedKeys::get,
-                     publisherDecorator(completedHandler, lostSegments -> {
+               // This covers when an iterator invokes hasNext multiple times each time returning false
+               // Otherwise the second hasNext would block for the timeout since we may not ever see a topology update
+               if (complete) {
+                  break;
+               }
+               // This is only entered if we couldn't complete all segments. This is because we either had a node leave
```

Comment:
Is "this" the body of the if statement? If so, I think it would be clearer to but the comment in the body, or say something like "cacheTopology is non-null only if the previous iteration finished without completing all the segments"

Reply:
I moved it inside the if block.